### PR TITLE
Move finalizing elemental UO after finalizing internal side UO

### PR DESCRIPTION
**Pull request by YaqiWang**
I am moving elemental UO finalize after internal side UO finalize. This does not break anything which means we do not have any dependency from side UO and internal side UO to elemental UO. Be warned that after this change I will create a dependency from side UO and internal side UO to elemental UO in Yak.

See #5972.

### CODE REVIEW

**Comment by permcody** [comment]
I'm OK with this but again I'd appreciate a dummy test that uses this dependency in MOOSE. It makes our life so much easier to test the things we rely on instead of waiting for applications to break. It should be trivial to make a dummy object that couples to the other one.

**Comment by YaqiWang** [comment]
Good suggestion, will do.

**Comment by YaqiWang** [comment]
I will just add one test. You guys might want add more tests to fix the ordering. @permcody Ready for your review.

**Comment by friedmud** [comment]
I'm fairly against this. Our current model makes no guarantees about
ordering between element and side UO execution. No one should be relying on
that ordering: one way or the other.

If we want ordering between these two then we need to do it properly, in
two sets of execution...

My point here is that you might want it one way but someone else might want
it the opposite. Neither person is right... so how do we choose? Do you win
because you were the first person to request this?

Neither one logically makes more sense either. I can see scenarios both
ways.

In summary: this is not a thing to change... because it's not something you
should be depending on in the first place.

**Comment by YaqiWang** [comment]
This is not the full dependency but rather who is got finalized first. To me, element UO always contains richer information and that is way I choose it to relay the information from side UOs. I did this because it will for sure make my life so much easier without impacting others. I am fine to take out the test. Ideally, we can do finalization smarter. So I think this PR should be OK.

**Comment by permcody** [comment]
While I'm against this as well I'm not sure I want to move to multiple groups at this time so this feels like the lesser of two evils. I do agree that at a framework level there really shouldn't be some implied ordering because it is difficult to argue which one is more correct than another.

So... The reason I asked for a test is to remind us that someone is or will depend on the existing ordering as a feature. Is it right? No, is it damaging to the internals of MOOSE? I don't really think so.

As Yaqi figured out, even though we say we don't guarantee ordering of execution we've already created some dependency with the NodalNormals system so "you allow it" already won here and we DO have a dependency right in MOOSE whether we like it or not. I could revive the whole dependency resolution system idea but I doubt anyone will get to that anytime soon.

If developers are willing to help us maintain interfaces and even test them. I think it's ok to make small concessions like this from time to time.

**Comment by andrsd** [comment]
> we've already created some dependency with the NodalNormals system.

This is __not__ a correct statement. The order of execution for user objects is given. It is: elemental, side, internal side, nodal, general. The nodal normals system was fit into this and did not create any extra dependency.

**Comment by permcody** [comment]
OK - fair enough but then your statement:

> The order of execution for user objects is given. It is: elemental, side, internal side, nodal, general. The nodal normals system was fit into this and did not create any extra dependency.

and @friedmud's statement are in conflict with one and other:

> Our current model makes no guarantees about
ordering between element and side UO execution. No one should be relying on
that ordering: one way or the other.

I believe our intention was to not make any guarantees but if they are always the same, then it's not unreasonable to find out that something relies on that order being the same.

**Comment by friedmud** [comment]
Right - I want there to be no guarantee on this at all.  Like... we should put a random number generator in there that finalizes them in a different order every time... just so people don't depend on it ;-)

BUT: Practicality overrules me for now ;-)

**Comment by andrsd** [comment]
> then your statement [...] and @friedmud's statement are in conflict with one and other:

Yes, I do not see a problem there :smiling_imp: So, when are we implementing the DAG, again?

**Comment by permcody** [comment]
I'll start working on the DAG when we model Fusion reactors... So I'd say about 10 years...

**Comment by andrsd** [comment]
Sounds like a plan...
